crates/rust/src/lib.rs

```diff
@@ -397,23 +493,208 @@ impl InterfaceGenerator<'_> {
+                        name.to_owned()
+                    };
+                    let impl_name = self
+                        .gen
+                        .opts
+                        .resource_exports
+                        .get(&path)
+                        .cloned()
+                        .or_else(|| self.gen.opts.stubs.then(|| "Stub".to_owned()))
+                        .ok_or_else(|| {
+                            format!("`resource_exports` parameter required for `{path}`")
+                        })?;
+
+                    uwriteln!(
+                        self.src,
+                        "use {path_to_root}{impl_name} as Rep{trait_name};"
+                    );
+                } else {
+                    let impl_name = impl_name.map_err(|s| s.clone())?;
+                    uwriteln!(
+                        self.src,
+                        "use {path_to_root}{impl_name} as {trait_name}Impl;"
+                    );
+                }
+                if saw_export {
+                    self.src.push_str("const _: () = {\n");
+                    for &func in &funcs {
+                        self.generate_guest_export(func, interface_name, &trait_name);
+                    }
+                    self.src.push_str("};\n");
+                }
+
+                if let Some(ty) = resource {
+                    self.finish_resource_export(ty);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn generate_imports<'a>(&mut self, funcs: impl Iterator<Item = &'a Function>) {
+        let wasm_import_module = self.wasm_import_module.unwrap();
+        let mut by_resource = group_by_resource(funcs);
+
+        // Make sure we generate code for resources with no methods:
+        match self.identifier {
+            Identifier::Interface(id, _) => {
+                for ty in self.resolve.interfaces[id].types.values() {
+                    if let TypeDefKind::Resource = &self.resolve.types[*ty].kind {
+                        by_resource.entry(Some(*ty)).or_default();
+                    }
+                }
+            }
+            Identifier::World(id) => {
+                let world = &self.resolve.worlds[id];
+                for item in world.imports.values() {
+                    if let WorldItem::Type(ty) = item {
+                        if let TypeDefKind::Resource = &self.resolve.types[*ty].kind {
+                            by_resource.entry(Some(*ty)).or_default();
+                        }
+                    }
+                }
+            }
+        }
+
+        for (resource, funcs) in by_resource {
+            if let Some(resource) = resource {
+                let name = self.resolve.types[resource].name.as_deref().unwrap();
+
+                let camel = name.to_upper_camel_case();
+
+                uwriteln!(
+                    self.src,
+                    r#"
+                        #[derive(Debug)]
+                        pub struct {camel} {{
```

Comment:
This is sort of similar to the C generator I think, but I'd ideally expect this to be present during the type declaration of the resource type itself. Will that not work out though?

Reply:
Ah sorry yeah those bits will need to happen at the end, but type declarations I'd hope could happen during the `type_foo` functions

Reply:
I could certainly do that.  The other reason I did it this way was to keep the type declaration for each resource close to its trait declaration, but I guess it would be nice to have all the types declared together.

Reply:
Yeah it's not critical really, although I think it's more important for C where items must be lexically "sorted" rather than the Rust style "dear rustc please figure it out"

Reply:
It already dawned on me that lexical sorting might be the only reasonable order for C++ to define types from more complex wit structures (I need to generate both host and guest side).

The combination of a method receiving/returning a typedef to a record and a record containing a typedef and a resource is quite common, so I didn't see any natural ordering for these header parts.

Reply:
Hm I'm not sure I quite understand, although @cpetig it sounds like you're saying that some patterns are going to be difficult to implement in C++? Could you sketch out a `*.wit` and hypothetical header with bindings "generated" to help me better understand the issue?

Reply:
Sorry, I tried to support your assessment that lexical sorting will be needed by a C or C++ code generator and might have replied a bit too terse. I am currently developing above mentioned C++ generators in https://github.com/cpetig/wit-bindgen/tree/resources/autosar and that directory contains two wit files which show this problem:

ErrorDomain (resource) depends on ErrorCodeType (typedef), ErrorCode (record) depends on the first resource and is aliased a lot (typedef), e.g. any of the future-X resources depend on the ErrorCode alias in radar.

[I am perfectly aware that these generators need re-work as I created them as I was exploring the code base]